File: examples/algorithms-go/leetcode/lc0232_implement_queue_using_stacks.go
package leetcode

// LeetCode 232. Implement Queue using Stacks
// 用栈实现队列
//
// Implement a first in first out (FIFO) queue using only two stacks.

// MyQueue 双栈法实现队列 FIFO 语义
// Two-stack approach for FIFO semantics
type MyQueue struct {
	inStack  []int
	outStack []int
}

// ConstructorQueue initializes the queue object
func ConstructorQueue() MyQueue {
	return MyQueue{inStack: []int{}, outStack: []int{}}
}

// Push pushes element x to the back of queue
func (this *MyQueue) Push(x int) {
	this.inStack = append(this.inStack, x)
}

// Pop removes the element from in front of queue and returns that element
// Amortized O(1)
func (this *MyQueue) Pop() int {
	this.Peek()
	val := this.outStack[len(this.outStack)-1]
	this.outStack = this.outStack[:len(this.outStack)-1]
	return val
}

// Peek gets the front element
// Transfer only when outStack is empty
func (this *MyQueue) Peek() int {
	if len(this.outStack) == 0 {
		for len(this.inStack) > 0 {
			val := this.inStack[len(this.inStack)-1]
			this.inStack = this.inStack[:len(this.inStack)-1]
			this.outStack = append(this.outStack, val)
		}
	}
	return this.outStack[len(this.outStack)-1]
}

// Empty returns whether the queue is empty
func (this *MyQueue) Empty() bool {
	return len(this.inStack) == 0 && len(this.outStack) == 0
}

File: examples/algorithms-rust/src/leetcode/lc0232_implement_queue_using_stacks.rs
//! LeetCode 232. Implement Queue using Stacks
//! 用栈实现队列
//!
//! Implement a first in first out (FIFO) queue using only two stacks.

pub struct MyQueue {
    in_stack: Vec<i32>,
    out_stack: Vec<i32>,
}

impl MyQueue {
    /// Initialize the queue object
    pub fn new() -> Self {
        MyQueue {
            in_stack: Vec::new(),
            out_stack: Vec::new(),
        }
    }

    /// Push element x to the back of queue
    pub fn push(&mut self, x: i32) {
        self.in_stack.push(x);
    }

    /// Removes the element from in front of queue and returns that element
    /// Amortized O(1)
    pub fn pop(&mut self) -> i32 {
        self.peek();
        self.out_stack.pop().unwrap()
    }

    /// Get the front element
    /// Transfer only when out_stack is empty
    pub fn peek(&mut self) -> i32 {
        if self.out_stack.is_empty() {
            while let Some(x) = self.in_stack.pop() {
                self.out_stack.push(x);
            }
        }
        *self.out_stack.last().unwrap()
    }

    /// Returns whether the queue is empty
    pub fn empty(&self) -> bool {
        self.in_stack.is_empty() && self.out_stack.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_queue_using_stacks() {
        let mut q = MyQueue::new();
        q.push(1);
        q.push(2);
        assert_eq!(q.peek(), 1);
        assert_eq!(q.pop(), 1);
        assert!(!q.empty());
        assert_eq!(q.pop(), 2);
        assert!(q.empty());
    }
}
